src/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/QueueItem.cs

```diff
@@ -56,59 +52,85 @@ internal class QueueItem<TRequest, TResponse, TRequestContext> : IQueueItem<TReq
         cancellationToken.Register(() => _completionSource.TrySetCanceled(cancellationToken));
 
         _logger = logger;
-        _request = request;
+        SerializedRequest = serializedRequest;
         LspServices = lspServices;
 
         MethodName = methodName;
-        Language = language;
 
         var telemetryService = lspServices.GetService<AbstractTelemetryService>();
 
         _requestTelemetryScope = telemetryService?.CreateRequestScope(methodName);
     }
 
-    public static (IQueueItem<TRequestContext>, Task<TResponse>) Create(
+    public static (IQueueItem<TRequestContext>, Task<object?>) Create(
         string methodName,
-        string language,
-        TRequest request,
+        object? serializedRequest,
         ILspServices lspServices,
         ILspLogger logger,
         CancellationToken cancellationToken)
     {
-        var queueItem = new QueueItem<TRequest, TResponse, TRequestContext>(
+        var queueItem = new QueueItem<TRequestContext>(
             methodName,
-            language,
-            request,
+            serializedRequest,
             lspServices,
             logger,
             cancellationToken);
 
         return (queueItem, queueItem._completionSource.Task);
     }
 
-    public async Task<TRequestContext> CreateRequestContextAsync(IMethodHandler handler, CancellationToken cancellationToken)
+    public async Task<TRequestContext> CreateRequestContextAsync<TRequest>(TRequest request, IMethodHandler handler, CancellationToken cancellationToken)
     {
         cancellationToken.ThrowIfCancellationRequested();
 
         _requestTelemetryScope?.RecordExecutionStart();
 
         var requestContextFactory = LspServices.GetRequiredService<AbstractRequestContextFactory<TRequestContext>>();
-        var context = await requestContextFactory.CreateRequestContextAsync(this, handler, _request, cancellationToken).ConfigureAwait(false);
+        var context = await requestContextFactory.CreateRequestContextAsync(this, handler, request, cancellationToken).ConfigureAwait(false);
         return context;
     }
 
+    public TRequest? TryDeserializeRequest<TRequest>(AbstractLanguageServer<TRequestContext> languageServer, RequestHandlerMetadata requestHandlerMetadata, bool isMutating)
+    {
+        try
+        {
+            return languageServer.DeserializeRequest<TRequest>(SerializedRequest, requestHandlerMetadata);
+        }
+        catch (Exception ex)
+        {
+            // Report the exception to logs / telemetry
+            _requestTelemetryScope?.RecordException(ex);
+            _logger.LogException(ex);
+
+            // Set the task result to the exception
+            _completionSource.TrySetException(ex);
```

Comment:
interesting htat we set the exception like this in the failure case.  but the above `try` doesn't do `TrySetResult` in the success case.

Also... this code will run in the case of cancellation.  Is that desirable?  Should cancellation be handled in this exact same way.  If so, can you add a comment to the effect that we *do* want that.

note: TrySetException on cancellation is odd.  I would expect TrySetCanceled.

Reply:
The `try` doesn't set a result because it needs to do more processing to set the actual result (invoke the handler).

However if we encounter an exception, we can't continue processing so we set the result on the source so the client can observe it (the request is failed).

> Also... this code will run in the case of cancellation. Is that desirable? Should cancellation be handled in this exact same way. If so, can you add a comment to the effect that we do want that.

Good question.  If there is no cancellation token passed here do we need to handle that?

Reply:
> Good question. If there is no cancellation token passed here do we need to handle that?

probably not.  but it would be good to then add asserst taht we do NOT expect to get cancellation exceptions and whatnot.

it's a little unclear to me how this all flows.  For example, some people pass the CT along as a property of some context-object.  I don't like that pattern because then i can't see if cancellation flows in.  But i really don't have a good sense of what is going on here.

Reply:
> The try doesn't set a result because it needs to do more processing to set the actual result (invoke the handler).

doc the code :)

Reply:
some more comments on this.  will share in person though.